tests/integration/cmd_devfile_deploy_test.go

```diff
@@ -650,4 +654,167 @@ CMD ["npm", "start"]
 			})
 		})
 	})
+
+	// More details on https://github.com/devfile/api/issues/852#issuecomment-1211928487
+	When("starting with Devfile with autoBuild or deployByDefault components", func() {
+		BeforeEach(func() {
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-autobuild-deploybydefault.yaml"),
+				filepath.Join(commonVar.Context, "devfile.yaml"),
+				helper.DevfileMetadataNameSetter(cmpName))
+		})
+
+		When("running odo deploy with some components not referenced in the Devfile", func() {
+			var stdout string
+
+			BeforeEach(func() {
+				stdout = helper.Cmd("odo", "deploy").AddEnv("PODMAN_CMD=echo").ShouldPass().Out()
+			})
+
+			It("should create the appropriate resources", func() {
+				By("automatically applying Kubernetes/OpenShift components with deployByDefault=true", func() {
+					for _, l := range []string{
+						"k8s-deploybydefault-true-and-referenced",
+						"k8s-deploybydefault-true-and-not-referenced",
```

Comment:
> On the main branch, if a component is unreferenced, we do not deploy it with `odo deploy`, this behavior is only seen on `odo dev`. However that behavior seems to have changed with this PR. I have only tested for deployByDefault=notset, but I assume it would work similarly for `deployByDefault=true`. Is this intended?
> 
Yes, that's intended. That's how I understood it, by making the rules explicitly documented and avoiding having different behaviors between the different commands.
I had added a comment related to that (it was most specifically about `odo deploy` potentially applying ServiceBinding Kubernetes components, since they are not referenced and have no `deployByDefault`): https://github.com/redhat-developer/odo/issues/5694#issuecomment-1464046208

> devfile.yaml
> ```shell
> ➜  PODMAN_CMD=echo odo deploy
>   __
>  /  \__     Running the application in Deploy mode using nodejs-prj1-api-abhz Devfile
>  \__/  \    Namespace: default
>  /  \__/    odo version: v3.8.0
>  \__/
> 
>  ⚠  You are using "default" namespace, odo may not work as expected in the default namespace.
>  ⚠  You may set a new namespace by running `odo create namespace <name>`, or set an existing one by running `odo set namespace <name>`
> 
> ↪ Deploying Kubernetes Component: my-component-unreferenced
>  ✓  Creating resource Deployment/my-component-unreferenced
>  ⚠  WARNING: Building images on Apple Silicon / M1 is not (yet) supported natively on Podman
>  ⚠  There is however a temporary workaround: https://github.com/containers/podman/discussions/12899
> 
> ↪ Building & Pushing Image: quay.io/unknown-account/myimage
>  •  Building image locally  ...
> build -t quay.io/unknown-account/myimage -f /tmp/nodejs-debug-2/Dockerfile /tmp/nodejs-debug-2
>  ✓  Building image locally [3ms]
>  •  Pushing image to container registry  ...
> push quay.io/unknown-account/myimage
>  ✓  Pushing image to container registry [2ms]
> 
> ↪ Deploying Kubernetes Component: my-component
>  ✓  Creating resource Deployment/my-component
> 
> Your Devfile has been successfully deployed
> ➜  k get deploy
> NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
> my-component                0/1     1            0           5s
> my-component-unreferenced   0/1     1            0           5s
> my-go                       0/1     1            0           11d
> ```
> 
> I wonder if the unreferenced component must be created after image build, just in case it wants to use the recently built image.

In your example, the image is referenced, so it will be applied when the `apply` command is invoked. Indeed, for resources that need to be applied automatically, this PR starts with image components first.

Reply:
>Yes, that's intended. That's how I understood it, by making the rules explicitly documented and avoiding having different behaviors between the different commands.

Ah alright, I missed that comment.

>Indeed, for resources that need to be applied automatically, this PR starts with image components first.

IIUC, you mean the unreferenced k8s resources with deployByDefault:unset/true will be created after the image components are built? or the other way around?

My point was, any k8s/oc resource regardless of it's reference/deployByDefault status should be created after the image is built.

Reply:
> > Yes, that's intended. That's how I understood it, by making the rules explicitly documented and avoiding having different behaviors between the different commands.
> 
> Ah alright, I missed that comment.
> 
> > Indeed, for resources that need to be applied automatically, this PR starts with image components first.
> 
> IIUC, you mean the unreferenced k8s resources with deployByDefault:unset/true will be created after the image components are built? or the other way around?
> 
Yes, the first way, i.e., we always start by applying image components that need to be created automatically, and then the rest.

> My point was, any k8s/oc resource regardless of it's reference/deployByDefault status should be created after the image is built.
Ah I see. That would be interesting indeed but could be a separate issue. But I think this can already be done via a composite command if you need some dependency between image and K8s/oc components, no?